frame/registry/src/lib.rs

```diff
@@ -683,25 +682,56 @@ pub mod pallet {
 			editable_by_registrar: bool,
 		) -> DispatchResult {
 			let required_deposit = Self::calculate_deposit(&data);
-			let old_deposit = Self::requests(&registry_index, &entity_id)
-				.map(|Request { deposit, .. }| deposit)
-				.unwrap_or_default();
 
 			// refund old deposit, if any
-			T::Currency::unreserve_named(T::ReserveIdentifier::get(), &owner, old_deposit);
+			if let Some(Some(req)) = Self::requests(&registry_index, &entity_id) {
+				T::Currency::unreserve_named(T::ReserveIdentifier::get(), &owner, req.deposit);
+			}
 
 			// reserve new deposit
 			T::Currency::reserve_named(T::ReserveIdentifier::get(), &owner, required_deposit)?;
 
 			Requests::<T, I>::insert(
 				&registry_index,
 				&entity_id,
-				Request { deposit: required_deposit, data, editable_by_registrar },
+				Some(Request { deposit: required_deposit, data, editable_by_registrar }),
 			);
 
 			Self::deposit_event(Event::RegistrationRequested { registry_index, entity_id });
 
 			Ok(())
 		}
+
+		fn unregister_hard(
+			registry_index: &RegistryIndex,
+			entity_id: &T::EntityId,
+		) -> DispatchResult {
+			if let Some(Registration { deposit, .. }) = Self::registries(&registry_index, entity_id)
+			{
+				// refund deposit
+				let owner = Self::entity_owner(entity_id).ok_or(Error::<T, I>::InvalidEntity)?;
+				T::Currency::unreserve_named(T::ReserveIdentifier::get(), &owner, deposit);
+			} else {
+				return Err(Error::<T, I>::InvalidEntity.into());
+			}
+			Registries::<T, I>::remove(registry_index, entity_id);
+
+			Ok(())
+		}
+
+		fn unregister_soft(
+			registry_index: &RegistryIndex,
+			entity_id: &T::EntityId,
+		) -> DispatchResult {
+			if let Some(req) = Self::requests(registry_index, entity_id) {
+				if req.is_some() {
+					return Err(Error::<T, I>::NotRequestedToUnregister.into());
+				}
+			}
+			Requests::<T, I>::remove(registry_index, entity_id);
+			Registries::<T, I>::remove(registry_index, entity_id);
```

Comment:
```suggestion
			Self::unregister_hard(registry_index, entity_id)
```

Otherwise:
1. we have code duplication
2. we don't refund deposit
3. we don't check if the company is even registered

I believe this one change fixes all of these.

Also: this means that we're missing tests for checking if deposit is correct after unregister.